rpmio/rpmlog.c

```diff
@@ -374,44 +340,37 @@ static int rpmlogDefault(FILE *stdlog, rpmlogRec rec)
 /* FIX: rpmlogMsgPrefix[] may be NULL */
 static void dolog(struct rpmlogRec_s *rec, int saverec)
 {
-    static pthread_mutex_t serialize = PTHREAD_MUTEX_INITIALIZER;
-
+    static std::mutex serialize;
     int cbrc = RPMLOG_DEFAULT;
     int needexit = 0;
     FILE *clog = NULL;
     rpmlogCallbackData cbdata = NULL;
     rpmlogCallback cbfunc = NULL;
-    rpmlogCtx ctx = rpmlogCtxAcquire(saverec);
-
-    if (ctx == NULL)
-	return;
+    rpmlogCtx ctx = rpmlogCtxAcquire();
```

Comment:
We're not using a read lock here any more (only a write lock if `saverec` is true), yet we read the `ctx` fields below. But I guess concurrent access is now prevented by the `serialize` lock, right?

Reply:
Restored the former way, only we now always take a write lock there. You win some, lose some...

But that inspired me to do what our kernel friends pointed out some time ago: eliminate one round of locking on each and every entry to rpmlog() - rpmlogSetMask() requires a lock because it does two operations, but rpmlog() entry only wants to read, and just reading an integer is atomic. (this is a separate new commit in there now)

Reply:
> Good question. I think I managed to convince myself that this is safe somehow but it's not: the serialize mutex will protect that section of code, but in the meanwhile the context is unlocked and so the callback could be changed halfway through us reading it, oops.

Oh, indeed, I see it now (in the old version of the commit), too :sweat_smile:. I didn't quite realize *that* initially but somehow the lock being scoped to the `saverec` block only was enough to raise a red flag :smile:

> With the STL locking stuff, one can't choose the lock type as dynamically as with the low-level pthread stuff, and I think this is me just trying to wipe that under the carpet. Good thing you spotted it. I guess we'll just have to take a write lock here always.

Yep, the patch is now much clearer, thanks!

Reply:
> But that inspired me to do what our kernel friends pointed out some time ago: eliminate one round of locking on each and every entry to rpmlog() - rpmlogSetMask() requires a lock because it does two operations, but rpmlog() entry only wants to read, and just reading an integer is atomic. (this is a separate new commit in there now)

Nice find, too. Yeah, this also improves readability a tiny bit, using rpmlogSetMask() to, well, *get* a mask, was a bit confusing regardless :smile: